Asgard/Communications/Classes/CbusEventManager.cs

```diff
@@ -0,0 +1,246 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Asgard.Data;
+using Asgard.Data.Interfaces;
+using Microsoft.Extensions.Logging;
+
+namespace Asgard.Communications
+{
+    public class CbusEventManager :
+        ICbusEventManager,
+        IDisposable
+    {
+        #region Fields
+
+        private readonly ICbusMessenger cbusMessenger;
+
+        private readonly ILogger logger;
+
+        private readonly ConcurrentDictionary<CbusEventKey, bool?> cbusEventStates = new();
+        private readonly ConcurrentDictionary<CbusEventKey, CbusEventCallback> cbusEventCallbacks = new();
+        private readonly EmptyingConcurrentQueue<Exception> cbusEventExceptionQueue = new();
+
+        private readonly CancellationTokenSource cancellationTokenSource = new();
+
+        #endregion
+
+        #region Properties
+
+        protected CancellationToken CancellationToken => this.cancellationTokenSource.Token;
+
+        #endregion
+
+        #region Constructors
+
+        public CbusEventManager(ICbusMessenger cbusMessenger, ILogger logger)
+        {
+            this.cbusMessenger = cbusMessenger;
+            this.logger = logger;
+
+            Task.Run(() => ProcessQueue());
+
+            this.cbusMessenger.MessageReceived += CbusMessenger_MessageReceived;
+        }
+
+        #endregion
+
+        #region IDisposable support
+
+        public bool IsDisposed { get; private set; }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!this.IsDisposed)
+            {
+                if (disposing)
+                {
+                    this.cancellationTokenSource.Cancel();
+                    // TODO: dispose managed state (managed objects)
```

Comment:
The CTS should be disposed here, not just cancelled

Reply:
Good catch!